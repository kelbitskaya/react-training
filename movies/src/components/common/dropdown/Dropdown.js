import React, {useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import {useLocation} from "react-router";

export default function Dropdown(props) {
  const { options, updateMovies } = props;
  const [selectedItem, setSelectedItem] = useState(0);
  const [activeItem, setActiveItem] = useState(0);

  let location = useLocation();

  useEffect(() => {
    const params = new URLSearchParams(location.search);
    const sortBy = params.get('sortBy');
    const index = options.reduce((p,c,i) => {return c.title === sortBy ?i:p}, 0);
    setSelectedItem(index);
  });


  const toggleDropdown = () => {
    setActiveItem(!activeItem);
  };

  const handleClick = (index) => {
    setSelectedItem(index);
    setActiveItem(!activeItem);
    if (updateMovies) {
      updateMovies(options[index]);
    }
  };

  const renderOptions = () => {
    return options.map((item, index) => (
      <li
        role="presentation"
        key={item.id}
        datatype={item.type}
        className="dropdown-content__item"
        onClick={() => handleClick(index)}
        onKeyDown={() => handleClick(index)}
        type="button"
      >
        {item.title}
      </li>
    ));
  };

  return (
    <div className="dropdown">
      <button
        className={`dropbtn ${(activeItem ? 'dropbtn__open' : '')}`}
        type="button"
        onClick={() => toggleDropdown()}
      >
        {options[selectedItem].title}
      </button>
      <ul
        className={`dropdown-content ${(activeItem ? 'dropdown-content__shown' : '')}`}
      >
        {renderOptions()}
      </ul>
    </div>
  )
};

Dropdown.propTypes = {
  options: PropTypes.arrayOf(PropTypes.node).isRequired,
  updateMovies: PropTypes.func.isRequired,
};
