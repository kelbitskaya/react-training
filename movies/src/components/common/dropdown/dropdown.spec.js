import React from 'react';
import rerender from 'react-test-renderer';
import Dropdown from './Dropdown';
import routeData from 'react-router';

const mockLocation = {
  pathname: '/welcome',
  hash: '',
  search: '',
  state: ''
};

const props = {
  options: [
    {
      title: "",
      value: "",
      id: new Date(),
      handleClick: jest.fn()
    }
  ],
  updateMovies: jest.fn(),
  toggleDropdown : jest.fn(),
};

describe('Dropdown', () => {
  beforeEach(()=>{
    jest.spyOn(routeData, 'useLocation').mockReturnValue(mockLocation)
  });

  it('toggleDropdown ', () => {
    const component = shallow(<Dropdown  {...props}/>);
    component.toggleDropdown = jest.fn();
    const dropdown = component.find('.dropbtn');
    dropdown.simulate('click');
    expect(dropdown.hasClass("dropbtn__open")).toEqual(false);
  });

  it('handleClick with click', () => {
    const component = shallow(<Dropdown  {...props}/>);
    component.handleClick = jest.fn();
    const dropdown = component.find('.dropdown-content__item');
    dropdown.simulate('click');
    expect(dropdown.handleClick).toBeUndefined();
  });

  it('handleClick with key down', () => {
    const component = shallow(<Dropdown  {...props}/>);
    component.handleClick = jest.fn();
    const dropdown = component.find('.dropdown-content__item');
    dropdown.simulate('keydown');
    expect(dropdown.handleClick).toBeUndefined();
  });

  it('Dropdown', () => {
    const component = rerender.create(
      <Dropdown
        {...props}
      />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
