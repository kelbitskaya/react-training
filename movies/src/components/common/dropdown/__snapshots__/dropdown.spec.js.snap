// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Dropdown 1`] = `
<div
  className="dropdown"
>
  <button
    className="dropbtn "
    onClick={[Function]}
    type="button"
  >
    
  </button>
  <ul
    className="dropdown-content "
  >
    <li
      className="dropdown-content__item"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="presentation"
      type="button"
    >
      
    </li>
  </ul>
</div>
`;
